### ISSUE TITLE
[context] element-based API

### ISSUE THREAD

**Issue by trusktr** [webcomponents-cg/community-protocols#62]
It would be great to have an element-based API, because it would make the context pattern not only usable with any Custom Element authoring patterns whether as `function` or `class` (Lit, solid-element, Enhance, Lume, etc), but also usable in any non-custom-element frameworks (React/Vue/Svelte/Solid/Angular/etc).

I wrote all the details and examples here initially:

- https://github.com/open-wc/context-protocol/issues/25

But maybe it would be more useful as an update to the context protocol here? 

The context protocol here documents only the event-based pattern. Would it be useful to also standardize both a mixin pattern and an element-based pattern? (Or at least the element pattern, because that's as stadards aligned as we can get, offering flexibility across all custom- and non-custom-element frameworks)

**Comment by Westbrook** [comment]
Doesn't _not_ having an element-based API make it easier for non-custom element solutions to leverage this as it is just event dispatching and listening that can happen no matter what your solution for applying those to an application are?

Blessing a single implementation in the form of actual elements feels contrary to the "protocol" concept in that it then becomes a mini-library that would just be another dependency that projects feel forced to take on to "make things work". Outlining how your application could interoperate with any number of other tools feels like a better goal for a "community protocol" than placing the onus on an application to "do it with these tools", which is generally what we see in frameworks already.

If the solution were to "ship an element" or "ship a mixing", how would you see the broader community take on management that? The theory of a central protocol that any library/framework/base class/project can adopt internally also avoids finding a solid solution to that, but if there were good one, maybe it would be worth learning in that direction both with the Context Protocol and other concepts managed in this repo.

**Comment by trusktr** [comment]
> Doesn't _not_ having an element-based API make it easier for non-custom element solutions to leverage this as it is just event dispatching and listening that can happen no matter what your solution for applying those to an application are?

Yeah, but this won't replace the JS event API. This feature is just more aligned with DOM and frameworks that manipulate DOM (React, Vue, Svelte, etc), where in all DOM frameworks we write things declaratively:

```jsx
return <some-context value=${...}>
  <some-element />
</some-context>
```

Being able to write declaratively makes it more convenient in the myriad of DOM frameworks. And also in plain HTML:

```html
<some-context value="...some deserializable string value...">
  <some-element />
</some-context>
```

which would work for SSR solutions.

The JS protocol would still be there for anyone who needs it, and this idea would not replace it (maybe it would be implemented on top of it).

Besides, the DOM is the most standard way to interop everywhere, because note that any DOM elements can be instantiated with JS, so absolutely any JS library or framework can easily do this:

```js
const context = document.createElement('some-context')
context.value = 123

context.append( someAPIThatCreatesATree() )
```

**Comment by Westbrook** [comment]
I'm likely blind to the full benefits, particularly in that I would _expect_ something like this to be "implemented on to of" the Context Protocol, but that doesn't mean we shouldn't get more information on what you'd want to make happen here. We've got a [Proposal Template](https://github.com/webcomponents-cg/community-protocols/blob/main/proposal-template.md), and likely the best next step is to structure a PR for ["Proposal status"](https://github.com/webcomponents-cg/community-protocols#:~:text=Proposal-,%22Proposal%22%20status,-applies%20to%20just) of such an API for others to get a better idea as to what you're thinking here. Sound good?

**Comment by trusktr** [comment]
Yes, it could indeed be implemented on top of the existing context protocol. This would add the additional layer that makes it declarative (HTML instead of JavaScript, which then makes it compatible with all non-custom-element frameworks).

Plus with an HTML representation, it will then become compatible with SSR, Declarative Custom Elements, etc. for example imagine a PHP or Rust server can send a top-level context in the HTML payload by outputting HTML without having to output JavaScript code.

**Comment by Westbrook** [comment]
We just had a great discussion on https://github.com/webcomponents-cg/community-protocols/pull/67 at the December meeting. Would you be interested/available in submitting a PR for discussion at the January or February sessions?

**Comment by trusktr** [comment]
Hey, I would be interested!  For those sorts of time sensitive questions better to ping me on Discord :)

**Comment by rektide** [comment]
I'm heavily interested in this topic, in trying to get state out of js and onto the page where it's straightforwardly visible. That feels like the view-source web that best empowers users, that makes evident what is. If the context is on the page, it's readily read/write by users & extensions, where-as the main proposal of event bubbling is very resistant to user agency.

One concern I have with the sketches here is that each context provider assets a single key-value. Having built web apps where there are many dozens of bits of initial page context, the thought of having the real page content be nested deep inside a lot of context providers sounds not great.

```html
<body>
<context-provider name="theme" value="cthonic">
<context-provider name="season" value="spring">
<context-provider name="transit" value="bike">
<div id="app">...</div>
</context-provider>
</context-provider>
</context-provider>
</body>
```

(I highly suspect this next thought would be deeply hated for not being conventionally semantic but), ergonomically it sure feels like it would be nice to represent context with key-values transposed out of providers and into children. To have context apply to its parent's scope.

```html
<body>
<context name="theme" value="cthonic"/>
<context name="season" value="spring">
<context name="transit" value="bike">
<div id="app">...</div>
</body>
```

Anything could carry context via this representation, and there's no nesting.

**Comment by justinfagnani** [comment]
@rektide Your first and second code block have the exact same nesting structure.

**Comment by ziadkh0** [comment]
@justinfagnani At first glance I thought the same, but then I noticed the `<context>` in the second code block was self-closing. So I think they meant for `<context>` to be more like `<br>` (in the sense of not taking children).

**Comment by justinfagnani** [comment]
@ziadkh0 HTML doesn't have self-closing tags, only a few void tags and there aren't going to be any more added.

**Comment by rektide** [comment]
I've updated my examples to conform to the non-consensual reality imposed upon us.

**Comment by justinfagnani** [comment]
@rektide fyi, you could write such a custom element today. All it has to do is attach an event listener to its parent.

**Comment by justinfagnani** [comment]
To clarify on that last comment, and at the risk of repeating myself and @Westbrook, any context elements would use the events in the protocol to communicate. This means that anyone can publish interoperable provider and consumer elements already. People can publish provider elements that provide to their children, that provide to their parent's children, that provide one context or many.

So is there anything actually actionable with with this issue?
